openvpn/primitives/generate.go

```diff
@@ -0,0 +1,134 @@
+package primitives
+
+import (
+	"crypto/ecdsa"
+	"crypto/x509"
+	"fmt"
+	log "github.com/cihub/seelog"
+	"os"
+	"path/filepath"
+)
+
+// SecurityPrimitives describes security primitives
+type SecurityPrimitives struct {
+	directory        string
+	CACertPath       string
+	CAKeyPath        string
+	ServerCertPath   string
+	ServerKeyPath    string
+	CRLPEMPath       string
+	TLSCryptKeyPath  string
+	caBytes          []byte
+	caPrivateKey     *ecdsa.PrivateKey
+	serverCertBytes  []byte
+	serverPrivateKey *ecdsa.PrivateKey
+}
+
+func exists(path string) (bool, error) {
+	_, err := os.Stat(path)
+	if err == nil {
+		return true, nil
+	}
+	if os.IsNotExist(err) {
+		return false, nil
+	}
+	return true, err
+}
+
+func (sp *SecurityPrimitives) mkDir(dir string) {
+	if e, err := exists(dir); !e {
+		log.Debug("Creating dir (" + dir + ")")
+		os.Mkdir(dir, 0700)
+	} else {
+		log.Errorf("Error creating directory", err)
+	}
+}
+
+const (
+	genDir    = "certs"
+	caDir     = "ca"
+	serverDir = "server"
+	clientDir = "clients"
+	commonDir = "common"
+)
+
+func (sp *SecurityPrimitives) init() {
+	sp.cleanupDir()
+
+	sp.mkDir(sp.directory)
+	sp.mkDir(filepath.Join(sp.directory, caDir))
+	sp.mkDir(filepath.Join(sp.directory, serverDir))
+	sp.mkDir(filepath.Join(sp.directory, clientDir))
+	sp.mkDir(filepath.Join(sp.directory, commonDir))
+}
+
+// NewOpenVPNSecPrimitives returns pre-initialized SecurityPrimitives structure
+func NewOpenVPNSecPrimitives() *SecurityPrimitives {
+	return &SecurityPrimitives{
+		genDir,
+		filepath.Join(genDir, caDir, "ca.crt"),
+		filepath.Join(genDir, caDir, "ca.key"),
+		filepath.Join(genDir, serverDir, "server.crt"),
+		filepath.Join(genDir, serverDir, "server.key"),
+		filepath.Join("bin", "tls", "crl.pem"),
+		filepath.Join(genDir, commonDir, "ta.key"),
+		nil,
+		nil,
+		nil,
+		nil,
```

Comment:
Very long list - since all params are given unnamed here, it's very easy to mismatch them. Especially since most of params are filenames.
```go
&SecurityPrimitives{
  CACertPath: filepath.Join(genDir, caDir, "ca.crt"),
  CAKeyPath: filepath.Join(genDir, caDir, "ca.key"),
  ...
}
```